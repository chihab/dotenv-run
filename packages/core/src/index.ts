export * from "./env.js";
export * from "./options.js";
export * from "./root.js";
export * from "./build.js";

declare global {
  interface ImportMeta {
    env: ImportMetaEnv;
  }

  interface ImportMetaEnv {
    [key: string]: any;
  }
}
