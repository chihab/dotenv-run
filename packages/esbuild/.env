MY_API_BASE=https://dotenv-run.app