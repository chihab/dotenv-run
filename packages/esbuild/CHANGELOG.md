# @dotenv-run/esbuild

## 1.2.3

### Patch Changes

- feat(core): support sub-module level turbo.json
- Updated dependencies
  - @dotenv-run/core@1.2.3

## 1.2.2

### Patch Changes

- fix(core): filter out undefined env values
- Updated dependencies
  - @dotenv-run/core@1.2.2

## 1.2.1

### Patch Changes

- refactor(esbuild): usde esbuild define option
- Updated dependencies
  - @dotenv-run/core@1.2.1

## 1.2.0

### Minor Changes

- f63ecac: feat: add esbuild plugin

### Patch Changes

- Updated dependencies [f63ecac]
  - @dotenv-run/core@1.2.0
