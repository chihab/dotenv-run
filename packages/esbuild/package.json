{
  "name": "@dotenv-run/esbuild",
  "version": "1.4.1",
  "description": "Run your scripts with dotenv variables",
  "homepage": "https://github.com/chihab/dotenv-run",
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "dev": "tsc -w",
    "test": "vitest",
    "build": "concurrently npm:build:*",
    "build:esm": "tsc --module esnext --outDir dist/esm",
    "build:cjs": "tsc --module commonjs  --outDir dist/cjs",
    "build:types": "tsc --emitDeclarationOnly --declaration --declarationMap --module esnext --outDir dist"
  },
  "files": [
    "dist",
    "README.md"
  ],
  "keywords": [
    "dotenv",
    "run",
    "cli"
  ],
  "author": "Chihab Otmani <chihab@gmail.com>",
  "license": "ISC",
  "dependencies": {
    "@dotenv-run/core": "workspace:~1.3.5"
  },
  "peerDependencies": {
    "esbuild": "^0.21.0"
  },
  "devDependencies": {
    "@types/node": "^16.18.112",
    "esbuild": "^0.23.0"
  }
}
