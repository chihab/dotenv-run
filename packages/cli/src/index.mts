#!/usr/bin/env node

import { env, findRootPath } from "@dotenv-run/core";
import chalk from "chalk";
import minimist from "minimist";
import { run } from "./run.js";

const argv = minimist(process.argv.slice(2), {
  string: ["debug", "env", "file", "root", "prefix"],
  boolean: ["help", "unsecure", "override"],
  alias: {
    debug: "d",
    env: "e",
    file: "f",
    root: "r",
    prefix: "p",
    help: "h",
    override: "o",
    unsecure: "u",
  },
  default: {
    unsecure: false,
    help: false,
    override: false,
  },
});

function help() {
  console.log(`
  Usage: dotenv-run [options] -- <command>
  
  Options:
    -d, --debug [regexp]                    print debug information, display environment variables filtered by regexp if provided
    -e, --env  [environment]                environment to load (default: NODE_ENV)
    -f, --file [.env,.secrets,.env.api]     specific .env files to load (default: .env, incompatible with root and prefix options)
    -r, --root                              root directory to search for .env files, defaults to current working directory
    -p, --prefix [.env,.secrets,.env.api]   .env file prefixes to load (default: .env)
    -h, --help                              output usage information
    -o, --override                          override existing environment variables (default: false)
    -u, --unsecure                          display environment variables in debug output (default: false)
    
  Examples:
    dotenv-run -d
    dotenv-run -- npm start
    dotenv-run -r ../.. -p .env,.secrets -- npm start
    dotenv-run -f ../.env,../.env.api -- npm start
  `);
}

function parseList(input: string | string[]): string[] {
  return typeof input === "string" ? input.split(",") : input;
}

if (argv.h) {
  help();
} else {
  const cmd = argv._[0];
  const verbose = argv.hasOwnProperty("d");
  if (!verbose && !cmd) {
    help();
    process.exit(1);
  }
  if (argv.f && (argv.r || argv.p)) {
    console.log(chalk.red("option -f cannot be used with -r or -p"));
    console.log("use -f to specify specific .env files");
    console.log("\tdotenv-run -f ./.env,../../secrets");
    console.log(
      "use -r to specify the directory from which to search for files prefixed with -p down to cwd"
    );
    console.log("\tdotenv-run -r ../.. -p .env,.env.api");
    process.exit(1);
  }
  const root = argv.f ? undefined : argv.r ?? findRootPath();
  const files = argv.f ? parseList(argv.f) : parseList(argv.p);
  const cwd = argv.f ? undefined : process.cwd();
  env({
    environment: argv.e ?? process.env.NODE_ENV,
    cwd,
    root,
    files,
    prefix: argv.d,
    verbose,
    unsecure: argv.u,
    dotenv: { override: argv.o },
  });
  if (cmd) run(cmd, argv._.slice(1));
}
