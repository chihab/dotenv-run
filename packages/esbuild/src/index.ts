import { Dict, env, type DotenvRunOptions } from "@dotenv-run/core";
import type { Plugin, PluginBuild } from "esbuild";

function definePluginBuild(build: PluginBuild, env: Dict): void {
  const define = build.initialOptions.define ?? {};
  build.initialOptions.define = {
    ...env,
    ...define,
  };
}

const dotenvRun = (options: DotenvRunOptions): Plugin => {
  return {
    name: "dotenv-run",
    setup(build) {
      const full = env(options).full;
      definePluginBuild(build, full);
    },
  };
};

const dotenvRunDefine = (env: Dict): Plugin => {
  return {
    name: "dotenv-run-define",
    setup(build) {
      definePluginBuild(build, env);
    },
  };
};

export { DotenvRunOptions, dotenvRun, dotenvRunDefine };
export default dotenvRun;
