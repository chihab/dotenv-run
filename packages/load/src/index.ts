import { env, findRootPath } from "@dotenv-run/core";
env({
  environment: process.env.NODE_ENV,
  cwd: process.cwd(),
  root: findRootPath(),
  files: [".env"]
});
env();
