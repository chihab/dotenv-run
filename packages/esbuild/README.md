# @dotenv-run/esbuild

- ✅ Load environment variables from the command line `API_BASE=/v1/ node esbuild.mjs`
- ✅ Load environment variables from `.env` files
- ✅ Expand environment variables `API_URL=$API_BASE/users`
- ✅ Define environment variables for a specific environment (e.g. `.env.production`)
- ✅ Load priorities of `.env.*` files (e.g. `.env.production` > `.env`)
- ✅ Hierarchical cascading configuration in monorepo projects ([Nx](https://nx.dev), [Turbo](https://turbo.build/), etc.)
  `apps/next-app/.env` > `apps/.env` > `.env`

## Install

```console
npm add @dotenv-run/esbuild --save-dev
```

## Usage

### Using define (recommended)

```js
import { build } from "esbuild";
import { env } from "@dotenv-run/esbuild";

const { full } = env({
  prefix: "MY_",
  verbose: true,
});

await build({
  bundle: true,
  write: false,
  entryPoints: [`test/app.js`],
  define: full,
});
```

### Using esbuild plugin

```js
import { build } from "esbuild";
import { env } from "@dotenv-run/esbuild";

const { full } = env({
  prefix: "MY_",
  verbose: true,
});

await build({
  bundle: true,
  write: false,
  entryPoints: [`test/app.js`],
  define: full,
});
```
