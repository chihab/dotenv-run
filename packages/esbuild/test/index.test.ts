import { build } from "esbuild";
import { describe, expect, it } from "vitest";
import { dotenvRun } from "../src/";
import { env } from "@dotenv-run/core";

describe("Usage with esbuild", () => {
  it("should replace environment variables using esbuild plugin", async () => {
    const results = await build({
      write: false,
      bundle: true,
      entryPoints: [`test/app.js`],
      plugins: [
        dotenvRun({
          prefix: "MY_",
          verbose: false,
          files: [".env"],
        }),
      ],
    });

    expect(results.outputFiles.at(0)?.text).toMatchSnapshot();
  });


  it("should replace environment variables using esbuild plugin and a custom define", async () => {
    const results = await build({
      write: false,
      bundle: true,
      entryPoints: [`test/app.js`],
      define: {
        "process.env.MY_API_BASE": '"https://dotenv-run.define"',
        "import.meta.env.MY_API_BASE": '"https://dotenv-run.define"',
      },
      plugins: [
        dotenvRun({
          prefix: "MY_",
          verbose: false,
          files: [".env"],
        }),
      ],
    });

    expect(results.outputFiles.at(0)?.text).toMatchSnapshot();
  });

  it("should replace environment variables using define option", async () => {
    const { full } = env({
      prefix: "MY_",
      verbose: false,
      files: [".env"],
    });

    const results = await build({
      bundle: true,
      write: false,
      entryPoints: [`test/app.js`],
      define: full,
    });

    expect(results.outputFiles.at(0)?.text.trim()).toMatchSnapshot();
  });
});
