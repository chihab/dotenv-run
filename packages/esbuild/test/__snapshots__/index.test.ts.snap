// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Usage with esbuild > should replace environment variables using define option 1`] = `
"\\"use strict\\";
(() => {
  // test/util.js
  function meta() {
    return \\"https://dotenv-run.app\\";
  }

  // test/app.js
  console.log(\\"https://dotenv-run.app\\");
  console.log(meta());
})();"
`;

exports[`Usage with esbuild > should replace environment variables using esbuild plugin 1`] = `
"\\"use strict\\";
(() => {
  // test/util.js
  function meta() {
    return \\"https://dotenv-run.app\\";
  }

  // test/app.js
  console.log(\\"https://dotenv-run.app\\");
  console.log(meta());
})();
"
`;

exports[`Usage with esbuild > should replace environment variables using esbuild plugin and a custom define 1`] = `
"\\"use strict\\";
(() => {
  // test/util.js
  function meta() {
    return \\"https://dotenv-run.define\\";
  }

  // test/app.js
  console.log(\\"https://dotenv-run.define\\");
  console.log(meta());
})();
"
`;
